### PR TITLE
{Network} `az network watcher connection-monitor`: Migrate to CodeGen V2

### PR DESCRIPTION
**Pull request by AllyW**
**Related command**

az network watcher connection-monitor create
az network watcher connection-monitor endpoint
az network watcher connection-monitor test-configuration
az network watcher connection-monitor test-group
az network watcher connection-monitor output

**Description**

**Testing Guide**

**History Notes**

[network] `az network watcher connection-monitor output`: Add, List and Remove

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Network

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 1]
```diff
@@ -7753,3 +7753,15 @@ def _output(self, *args, **kwargs):
...
+def remove_nw_connection_monitor_output(cmd, connection_monitor_name, location):
+
+    update_args = {
+        'connection_monitor_name': connection_monitor_name,
+        'location': location
+    }
+    from azure.cli.core.commands import LongRunningOperation
+    from .operations.watcher import WatcherConnectionMonitorOutputRemove
+    poller = WatcherConnectionMonitorOutputRemove(cli_ctx=cmd.cli_ctx)(command_args=update_args)
+    return LongRunningOperation(cmd.cli_ctx)(poller)
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:7767]: You can directly return the poller

```suggestion
    return WatcherConnectionMonitorOutputRemove(cli_ctx=cmd.cli_ctx)(command_args=update_args)
```

File: src/azure-cli/azure/cli/command_modules/network/operations/watcher.py [hunk 2]
```diff
@@ -378,7 +366,89 @@ def _build_arguments_schema(cls, *args, **kwargs):
...
+    @classmethod
+    def _build_arguments_schema(cls, *args, **kwargs):
+        args_schema = super()._build_arguments_schema(*args, **kwargs)
+        args_schema.watcher_name._registered = False
+        args_schema.watcher_name._required = False
+        args_schema.watcher_rg._registered = False
+        args_schema.watcher_rg._required = False
+        args_schema.migrate._registered = False
+        args_schema.output_index._registered = False
+        args_schema.no_wait._registered = False
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/operations/watcher.py:396]: If the generated codegen add --no-wait arguments, we can expose it directly.

File: src/azure-cli/azure/cli/command_modules/network/operations/watcher.py [hunk 3]
```diff
@@ -378,7 +366,89 @@ def _build_arguments_schema(cls, *args, **kwargs):
...
+
+    @classmethod
+    def _build_arguments_schema(cls, *args, **kwargs):
+        args_schema = super()._build_arguments_schema(*args, **kwargs)
+        args_schema.watcher_name._registered = False
+        args_schema.watcher_name._required = False
+        args_schema.watcher_rg._registered = False
+        args_schema.watcher_rg._required = False
+        args_schema.migrate._registered = False
+        args_schema.output_index._registered = False
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/operations/watcher.py:395]: output_index can also be exposed to the users.

File: src/azure-cli/azure/cli/command_modules/network/operations/watcher.py [hunk 4]
```diff
@@ -378,7 +366,89 @@ def _build_arguments_schema(cls, *args, **kwargs):
...
+class WatcherConnectionMonitorOutputAdd(_WatcherConnectionMonitorOutputAdd):
+
+    @classmethod
+    def _build_arguments_schema(cls, *args, **kwargs):
+        args_schema = super()._build_arguments_schema(*args, **kwargs)
+        args_schema.watcher_name._registered = False
+        args_schema.watcher_name._required = False
+        args_schema.watcher_rg._registered = False
+        args_schema.watcher_rg._required = False
+        args_schema.migrate._registered = False
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/operations/watcher.py:394]: If migrate is not used, you should hide it in aaz-dev-tools

File: src/azure-cli/azure/cli/command_modules/network/operations/watcher.py [hunk 5]
```diff
@@ -378,7 +366,89 @@ def _build_arguments_schema(cls, *args, **kwargs):
...
+        get_network_watcher_from_location(self,
+                                          watcher_name='network_watcher_name',
+                                          rg_name='resource_group_name'
+                                          )
+
+
+def process_nw_cm_v2_output_namespace(cmd):
+    args = cmd.ctx.args
+    v2_output_optional_parameter_set = ['workspace_id']
+    if hasattr(args, 'out_type') and args.out_type.to_serialized_data() is not None:
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/operations/watcher.py:378]: Please use has_value(args.out_type) if use not define args.out_type. The value of arg can be `AAZUndefined` if the user does not define this argument or `None` of user want to clear this property. The `has_value` functino is used to detect `AAZUndefined`.

**Comment by kairu-ms** [comment]
Please record the related tests in live mode again.

**Review by kairu-ms — DISMISSED** [review]

File: src/azure-cli/azure/cli/command_modules/network/operations/watcher.py [hunk 6]
```diff
@@ -550,19 +564,334 @@ def pre_operations(self):
...
+        if not has_value(args.location):
+            raise ValidationError("Can not get location from --endpoint-source-resource-id")
+
+    v2_required_parameter_set = [
+        'endpoint_source_resource_id', 'endpoint_source_name', 'endpoint_dest_name', 'test_config_name'
+    ]
+    for p in v2_required_parameter_set:
+        if not hasattr(args, p) or getattr(args, p) is None:
+            raise ValidationError(
+                'usage error: --{} is required to create a V2 connection monitor'.format(p.replace('_', '-')))
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/operations/watcher.py:598]: Recommend directly using `args.endpoint_source_resource_id`

File: src/azure-cli/azure/cli/command_modules/network/operations/watcher.py [hunk 7]
```diff
@@ -1096,3 +1402,706 @@ def pre_operations(self):
         get_network_watcher_from_resource(self)
         if has_value(args.resource):
             args.target_resource_id = args.resource
+
+
+def process_nw_cm_v2_output_namespace(cmd):
+    args = cmd.ctx.args
+    v2_output_optional_parameter_set = ['workspace_id']
+    if hasattr(args, 'out_type') and args.out_type.to_serialized_data() is not None:
+        tmp = [p for p in v2_output_optional_parameter_set if getattr(args, p) is None]
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/operations/watcher.py:1411]: Same concern, not recommend using `getattr` or `setattr`

File: src/azure-cli/azure/cli/command_modules/network/operations/watcher.py [hunk 8]
```diff
@@ -1096,3 +1402,706 @@ def pre_operations(self):
...
+        get_network_watcher_from_location(self)
+
+        args.scope_exclude = []
+        args.scope_include = []
+        if has_value(args.address_include):
+            for ip in args.address_include:
+                args.scope_include.append({"address": ip})
+        if has_value(args.address_exclude):
+            for ip in args.address_exclude:
+                args.scope_exclude.append({"address": ip})
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/network/operations/watcher.py:1545]: Please use `assign_aaz_list_arg`

**Review by kairu-ms — APPROVED** [review]

**Pull request merged**